### PR TITLE
SLEEVE: FIX #4022, #4024, #4025, #3998

### LINKED ISSUE #4022: Bladeburner with negative contracts remaining
Bitburner v2.0.2 (3067703c)
[bitburnerSave_1661188743_BN12x26.zip](https://github.com/danielyxie/bitburner/files/9396335/bitburnerSave_1661188743_BN12x26.zip)

I set a sleeve to perform bladeburner actions, take on contracts doing tracking actions. In the past it would stop once remaining contracts was 0. Once it hit zero it flipped to negative number and kept going.

### PR DESCRIPTION
**Pull request by Mughur**
#4022: Probably not the best way of fixing the issue, as it still let's the sleeve start the action, but stops the action when `process` is called.

#4024: Hospitalizing goes down, but not to negative to prevent issues

#4025: UI now takes BN multipliers into account

#3998: Sleeve BB contracts now reward money

-sleeve.getTask() now returns the name of the crime being committed in `name` field

-Sleeve BB actions now give exp.

-Bladeburner's getActionStats now passes the person object so that action rewards can be completely calculated based on what the person is doing, while before the rewards `time` multiplier was using the value of the player's current action's time.

### CODE REVIEW

File: src/PersonObjects/Sleeve/Work/SleeveCrimeWork.ts [hunk 1]
```diff
@@ -64,6 +64,7 @@ export class SleeveCrimeWork extends Work {
   APICopy(): Record<string, unknown> {
     return {
       type: this.type,
+      name: this.crimeType
```
Comment (hydroflame) [inline src/PersonObjects/Sleeve/Work/SleeveCrimeWork.ts:67]: the player version is called `crimeType` so this should be too.

File: src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts [hunk 2]
```diff
@@ -34,16 +36,34 @@ export class SleeveBladeburnerWork extends Work {
...
+        }
+      }
+      const retValue = player.bladeburner.completeAction(player, sleeve, actionIdent, false);
       let exp: WorkStats | undefined;
       if (this.actionType === "General") {
         exp = GeneralActions[this.actionName]?.exp;
         if (!exp) throw new Error(`Somehow there was no exp for action ${this.actionType} ${this.actionName}`);
         applySleeveGains(player, sleeve, exp, 1);
       }
+      if (retValue.money>0)player.gainMoney(retValue.money,"sleeves")
```
Comment (hydroflame) [inline src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts:66]: fmt your code.

File: src/PersonObjects/Sleeve/ui/StatsElement.tsx [hunk 3]
```diff
@@ -100,51 +101,51 @@ export function EarningsElement(props: IProps): React.ReactElement {
...
-      [`Agility Exp:`, `${numeralWrapper.formatExp(5 * rates.agiExp)} / sec`],
-      [`Charisma Exp:`, `${numeralWrapper.formatExp(5 * rates.chaExp)} / sec`],
-      [`Reputation:`, <ReputationRate reputation={5 * repGain} />],
+      [`Hacking Exp:`, `${numeralWrapper.formatExp(5 * rates.hackExp * BitNodeMultipliers.FactionWorkExpGain)} / sec`],
+      [`Strength Exp:`, `${numeralWrapper.formatExp(5 * rates.strExp * BitNodeMultipliers.FactionWorkExpGain)} / sec`],
+      [`Defense Exp:`, `${numeralWrapper.formatExp(5 * rates.defExp * BitNodeMultipliers.FactionWorkExpGain)} / sec`],
+      [`Dexterity Exp:`, `${numeralWrapper.formatExp(5 * rates.dexExp * BitNodeMultipliers.FactionWorkExpGain)} / sec`],
+      [`Agility Exp:`, `${numeralWrapper.formatExp(5 * rates.agiExp * BitNodeMultipliers.FactionWorkExpGain)} / sec`],
+      [`Charisma Exp:`, `${numeralWrapper.formatExp(5 * rates.chaExp * BitNodeMultipliers.FactionWorkExpGain)} / sec`],
+      [`Reputation:`, <ReputationRate reputation={5 * repGain * BitNodeMultipliers.FactionWorkRepGain} />],
```
Comment (hydroflame) [inline src/PersonObjects/Sleeve/ui/StatsElement.tsx:135]: That doesn't seem right, Work/formulas/Company.ts already has the bitnode mults.

Comment (Mughur) [reply src/PersonObjects/Sleeve/ui/StatsElement.tsx]: I should've double checked them. Some of those calculations were correct, some of them were not.

File: src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts (hunk 2 above)
Comment (Mughur) [reply src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts]: done

File: src/PersonObjects/Sleeve/Work/SleeveCrimeWork.ts (hunk 1 above)
Comment (Mughur) [reply src/PersonObjects/Sleeve/Work/SleeveCrimeWork.ts]: switched

File: src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts [hunk 4]
```diff
@@ -34,16 +34,34 @@ export class SleeveBladeburnerWork extends Work {
...
+        }
+      }
+      const retValue = player.bladeburner.completeAction(player, sleeve, actionIdent, false);
       let exp: WorkStats | undefined;
       if (this.actionType === "General") {
         exp = GeneralActions[this.actionName]?.exp;
         if (!exp) throw new Error(`Somehow there was no exp for action ${this.actionType} ${this.actionName}`);
         applySleeveGains(player, sleeve, exp, 1);
       }
+      if (retValue.money > 0) player.gainMoney(retValue.money, "sleeves");
```
Comment (phyzical) [inline src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts:64]: i wonder if this should occur as part of sleeve gains?

File: src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts [hunk 5]
```diff
@@ -34,16 +34,34 @@ export class SleeveBladeburnerWork extends Work {
...
+    const actionIdent = player.bladeburner.getActionIdFromTypeAndName(this.actionType, this.actionName);
+    if (!actionIdent) throw new Error(`Error getting ${this.actionName} action`);
+    if (this.actionType === "Contracts") {
+      const action = player.bladeburner.getActionObject(actionIdent);
+      if (!action) throw new Error(`Error getting ${this.actionName} action object`);
+      if (action.count <= 0) {
+        sleeve.stopWork(player);
+        return 0;
+      }
+    }
```
Comment (phyzical) [inline src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts:46]: nit: make make this an anon func localy and call it here and in the while loop

File: src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts [hunk 6]
```diff
@@ -34,16 +34,34 @@ export class SleeveBladeburnerWork extends Work {
...
+          sleeve.stopWork(player);
+          return 0;
+        }
+      }
+      const retValue = player.bladeburner.completeAction(player, sleeve, actionIdent, false);
       let exp: WorkStats | undefined;
       if (this.actionType === "General") {
         exp = GeneralActions[this.actionName]?.exp;
         if (!exp) throw new Error(`Somehow there was no exp for action ${this.actionType} ${this.actionName}`);
         applySleeveGains(player, sleeve, exp, 1);
```
Comment (coderanger) [inline src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts:62]: Should this also check for the XP gains in `retValue` like happens with a Player?

File: src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts (hunk 4 above)
Comment (Mughur) [reply src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts:64]: Preferably not, but putting it elsewhere would require quite a bit of refactoring or passing the `retValue` around

File: src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts (hunk 5 above)
Comment (Mughur) [reply src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts:46]: Fair enough

File: src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts (hunk 6 above)
Comment (Mughur) [reply src/PersonObjects/Sleeve/Work/SleeveBladeburnerWork.ts:62]: yes, I had to do some more work to get the exp calculations to work tho

**Pull request merged**